src/plugins/chart_expressions/expression_xy/public/types.ts

```diff
@@ -0,0 +1,152 @@
+import { ChartsPluginSetup } from '../../../charts/public';
+import { IFieldFormat, SerializedFieldFormat } from '../../../../plugins/field_formats/common';
+import type { RangeSelectContext, ValueClickContext } from '../../../../plugins/embeddable/public';
+import { Datatable, ExpressionsServiceStart, ExpressionsSetup } from '../../../expressions/public';
+
+export interface SetupDeps {
+  expressions: ExpressionsSetup;
+  data: DataPublicPluginSetup;
+  fieldFormats: FieldFormatsSetup;
+  charts: ChartsPluginSetup;
+}
+
+export interface StartDeps {
+  expression: ExpressionsServiceStart;
+}
+
+export type ExpressionXyPluginSetup = void;
+export type ExpressionXyPluginStart = void;
+
+export interface FilterEvent {
+  name: 'filter';
+  data: ValueClickContext['data'];
+}
+
+export interface BrushEvent {
+  name: 'brush';
+  data: RangeSelectContext['data'];
+}
+
+export type FormatFactory = (mapping?: SerializedFieldFormat) => IFieldFormat;
+
+export interface OperationDescriptor extends Operation {
+  hasTimeShift: boolean;
+}
+
+export type SortingHint = 'version';
+export type FieldOnlyDataType = 'document' | 'ip' | 'histogram' | 'geo_point' | 'geo_shape';
+export type DataType = 'string' | 'number' | 'date' | 'boolean' | FieldOnlyDataType;
+
+export interface OperationMetadata {
+  // The output of this operation will have this data type
+  dataType: DataType;
+  // A bucketed operation is grouped by duplicate values, otherwise each row is
+  // treated as unique
+  isBucketed: boolean;
+  /**
+   * ordinal: Each name is a unique value, but the names are in sorted order, like "Top values"
+   * interval: Histogram data, like date or number histograms
+   * ratio: Most number data is rendered as a ratio that includes 0
+   */
+  scale?: 'ordinal' | 'interval' | 'ratio';
+  // Extra meta-information like cardinality, color
+  // TODO currently it's not possible to differentiate between a field from a raw
+  // document and an aggregated metric which might be handy in some cases. Once we
+  // introduce a raw document datasource, this should be considered here.
+  isStaticValue?: boolean;
+}
+
+export interface Operation extends OperationMetadata {
+  // User-facing label for the operation
+  label: string;
+  sortingHint?: SortingHint;
+}
+
+export interface FramePublicAPI {
+  datasourceLayers: Record<string, DatasourcePublicAPI>;
+  appliedDatasourceLayers?: Record<string, DatasourcePublicAPI>; // this is only set when auto-apply is turned off
+  /**
+   * Data of the chart currently rendered in the preview.
+   * This data might be not available (e.g. if the chart can't be rendered) or outdated and belonging to another chart.
+   * If accessing, make sure to check whether expected columns actually exist.
+   */
+  activeData?: Record<string, Datatable>;
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
```

Comment:
if you change the code with the comment above, that's not needed either.

Reply:
Addressed.